### PR TITLE
Optimize blendmap generation

### PR DESCRIPTION
**Pull request by akortunov**
Based on bzzt's changes.

Currently for some cells blendmap generation takes about 3ms, what can lead to FPS drop during cell transitions. With this PR generation takes about 0.5ms (roughly 6x faster).

The main idea - do not blend with black texture, get rid of outer loop.

### CODE REVIEW

**Comment by akortunov** [comment]
Any objections here?

File: components/esmterrain/storage.cpp [hunk 1]
```diff
@@ -385,72 +380,61 @@ namespace ESMTerrain
...
-        for (int i=0; i<numBlendmaps; ++i)
-        {
-            osg::ref_ptr<osg::Image> image (new osg::Image);
-            image->allocateImage(blendmapImageSize, blendmapImageSize, 1, GL_ALPHA, GL_UNSIGNED_BYTE);
-            unsigned char* pData = image->data();
+        LandCache cache;
+        std::map<UniqueTextureId, unsigned int> textureIndicesMap;
 
-            for (int y=0; y<blendmapSize; ++y)
+        for (int y=0; y<blendmapSize; ++y)
```
Comment (Capostrophic) [inline components/esmterrain/storage.cpp]: Starting to nitpick. Making integer copies (4 bytes) is cheaper on 64-bit architecture than making integer references (8 bytes).

File: components/esmterrain/storage.cpp [hunk 2]
```diff
@@ -385,72 +380,61 @@ namespace ESMTerrain
...
+                        if (layerList[i].mDiffuseMap == info.mDiffuseMap)
+                        {
+                            layerIndex = i;
+                            break;
+                        }
+                    }
 
-                    int realY = (blendmapSize - y - 1)*imageScaleFactor;
-                    int realX = x*imageScaleFactor;
+                    found = textureIndicesMap.insert(std::make_pair(id, layerIndex)).first;
```
Comment (Capostrophic) [inline components/esmterrain/storage.cpp]: emplace

File: components/esmterrain/storage.cpp [hunk 3]
```diff
@@ -385,72 +380,61 @@ namespace ESMTerrain
...
-                    pData[(realY+0)*blendmapImageSize + realX + 1] = alpha;
-                    pData[(realY+1)*blendmapImageSize + realX + 1] = alpha;
+                    if (layerIndex >= layerList.size())
+                    {
+                        osg::ref_ptr<osg::Image> image (new osg::Image);
+                        image->allocateImage(blendmapImageSize, blendmapImageSize, 1, GL_ALPHA, GL_UNSIGNED_BYTE);
+                        unsigned char* pData = image->data();
+                        memset(pData, 0, image->getTotalDataSize());
+                        blendmaps.push_back(image);
+                        layerList.push_back(info);
```
Comment (Capostrophic) [inline components/esmterrain/storage.cpp]: can use emplace_back here

File: components/esmterrain/storage.cpp [hunk 4]
```diff
@@ -385,72 +380,61 @@ namespace ESMTerrain
...
 
-                    int realY = (blendmapSize - y - 1)*imageScaleFactor;
-                    int realX = x*imageScaleFactor;
+                    found = textureIndicesMap.insert(std::make_pair(id, layerIndex)).first;
 
-                    pData[(realY+0)*blendmapImageSize + realX + 0] = alpha;
-                    pData[(realY+1)*blendmapImageSize + realX + 0] = alpha;
-                    pData[(realY+0)*blendmapImageSize + realX + 1] = alpha;
-                    pData[(realY+1)*blendmapImageSize + realX + 1] = alpha;
+                    if (layerIndex >= layerList.size())
```
Comment (Capostrophic) [inline components/esmterrain/storage.cpp]: layerIndex will never be greater than the size of the list. In fact, the whole layer index handling in this scope is obscure. Not sure how exactly to adjust the surrounding code with that in mind, though.

Comment (akortunov) [reply components/esmterrain/storage.cpp]: But it still can be **equal** to the size of list. It seems the > is here just for sure.

**Comment by Capostrophic** [comment]
I'll bite the bullet.

**Pull request merged**